provisioner/jwks.go

```diff
@@ -0,0 +1,142 @@
+package provisioner
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+
+	"github.com/pkg/errors"
+	"gopkg.in/square/go-jose.v2"
+)
+
+const (
+	kubernetesProgrammaticAuthorization = "urn:kubernetes:programmatic_authorization"
+	idTokenResponse                     = "id_token"
+	publicSubject                       = "public"
+	wellKnowOidcConfigurationUrl        = "https://%s.%s"
+	jwksUrl                             = "https://%s.%s/openid-configuration/keys.json"
+	subjectClaim                        = "sub"
+	issuerClaim                         = "iss"
+)
+
+// Provider contains the subset of the OpenID Connect provider metadata needed to request
+// and verify ID Tokens.
+type Provider struct {
+	Issuer                 string   `json:"issuer"`
+	AuthURL                string   `json:"authorization_endpoint"`
+	JWKSURL                string   `json:"jwks_uri"`
+	SupportedResponseTypes []string `json:"response_types_supported"`
+	SupportedSubjectTypes  []string `json:"subject_types_supported"`
+	AlgorithmsSupported    []string `json:"id_token_signing_alg_values_supported"`
+	SupportedClaims        []string `json:"claims_supported"`
+}
+
+// copied from kubernetes/kubernetes#78502
+func keyIDFromPublicKey(publicKey interface{}) (string, error) {
+	publicKeyDERBytes, err := x509.MarshalPKIXPublicKey(publicKey)
+	if err != nil {
+		return "", fmt.Errorf("failed to serialize public key to DER format: %v", err)
+	}
+
+	hasher := crypto.SHA256.New()
+	_, err = hasher.Write(publicKeyDERBytes)
+	if err != nil {
+		return "", fmt.Errorf("failed to write sha256 of public key: %v", err)
+	}
+	publicKeyDERHash := hasher.Sum(nil)
+
+	keyID := base64.RawURLEncoding.EncodeToString(publicKeyDERHash)
+
+	return keyID, nil
+}
+
+type KeyResponse struct {
+	Keys []map[string]string `json:"keys"`
+}
+
+func generateJWKSDocument(serviceAccountKey string) (string, error) {
+	block, _ := pem.Decode([]byte(serviceAccountKey))
+	if block == nil {
+		return "", errors.Errorf("Error decoding serviceAccountKey")
+	}
+
+	pubKey, err := x509.ParsePKIXPublicKey(block.Bytes)
+	if err != nil {
+		return "", fmt.Errorf("failed to parse public key: %v", err)
+	}
+	switch pubKey.(type) {
+	case *rsa.PublicKey:
+	default:
+		return "", errors.New("Public key was not RSA")
+	}
+
+	var alg jose.SignatureAlgorithm
+	switch pubKey.(type) {
+	case *rsa.PublicKey:
+		alg = jose.RS256
+	default:
+		return "", fmt.Errorf("invalid public key type %T, must be *rsa.PrivateKey", pubKey)
```

Comment:
this can't happen anymore, because in line 71-75 you already checked that

Reply:
fixed